src/zenml/integrations/gcp/model_registries/vertex_model_registry.py

```diff
@@ -0,0 +1,308 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Vertex AI model registry integration for ZenML."""
+
+from datetime import datetime
+from typing import Any, Dict, List, Optional, cast
+
+from google.cloud import aiplatform
+
+from zenml.integrations.gcp.flavors.vertex_model_registry_flavor import (
+    VertexAIModelRegistryConfig,
+)
+from zenml.integrations.gcp.google_credentials_mixin import (
+    GoogleCredentialsMixin,
+)
+from zenml.logger import get_logger
+from zenml.model_registries.base_model_registry import (
+    BaseModelRegistry,
+    ModelRegistryModelMetadata,
+    ModelVersionStage,
+    RegisteredModel,
+    RegistryModelVersion,
+)
+
+logger = get_logger(__name__)
+
+
+class VertexAIModelRegistry(BaseModelRegistry, GoogleCredentialsMixin):
+    """Register models using Vertex AI."""
+
+    @property
+    def config(self) -> VertexAIModelRegistryConfig:
+        """Returns the config of the model registry.
+
+        Returns:
+            The configuration.
+        """
+        return cast(VertexAIModelRegistryConfig, self._config)
+
+    def setup_aiplatform(self) -> None:
+        """Setup the Vertex AI platform."""
+        credentials, project_id = self._get_authentication()
+        aiplatform.init(
+            project=project_id,
+            location=self.config.location,
+            credentials=credentials,
+        )
```

Comment:
Just a note that calling this will authenticate the entire Vertex AI library globally. If there's some other code that calls the same function (like instantiating a Vertex orchestrator or Vertex model deployer), it will overwrite it, creating authentication errors that are very difficult to pinpoint and debug.

The right way to do this would be to create an individual client or session with its own credentials and attach it to this object, but clients are far more difficult to work with:

```
credentials = ...  # Load your custom credentials
project_id = "your-project-id"
location = "us-central1"

# Initialize the ModelServiceClient
client = aiplatform_v1.ModelServiceClient(credentials=credentials)

# Define the parent resource (project and location)
parent = f"projects/{project_id}/locations/{location}"

# List models with optional filtering
filter_expr = "display_name:my-model-name"  # Optional filter expression
models = client.list_models(parent=parent, filter=filter_expr)
```

Instead, I recommend you just check the other VertexAI stack components and make sure that they always re-initialize the global VertexAI session before they call any of its APIs.

Reply:
Update: you might be able to pass the credentials straight to the Model object, at least [according to the docs](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform.Model):

```
Model(
    model_name: str,
    project: typing.Optional[str] = None,
    location: typing.Optional[str] = None,
    credentials: typing.Optional[google.auth.credentials.Credentials] = None,
    version: typing.Optional[str] = None,
)```